### PR TITLE
Transformer Block (for HuggingFace support)

### LINKED ISSUE #721: Port TransformerBlock

### PR DESCRIPTION
**Pull request by sararb**
Fixes #721 

### Blocked by
- #768

### Goals :soccer:
- [x] Port TransformerBlock from transformers4rec
- [x] Implement specific HF transformer blocks: Bert, Albert, Roberta, XLNet, and GPT2

### Implementation Details :construction:
- Port the TransformerBlock from T4Rec and refactor it based on the latest HuggingFace code
- Add `pre` layer to support preparing additional inputs of the transformer model such as `attention_mask` or `head_mask`. This layer has access to the sequence of input embeddings and raw features. 
- Add a post with an `output_fn` function to process the [TFBaseModelOutputWithPoolingAndCrossAttentions](https://github.com/huggingface/transformers/blob/0b567aa430e9b1b56254de4775db5868104d34e5/src/transformers/modeling_tf_outputs.py#L125) return by HF layer and return the desired tensor(s). 
- Add custom transformers blocks (bert, albert..) with the `build_transformer_layer` method that prepares the HF configuration class with unified names of the required parameters `(d_model, n_head, n_layer,max_seq_length)`

### Testing Details :mag:
Add unit tests of the transformer block: 

- Test the TransformerBlock as a sequence encoder
- Test pooling the sequence of hidden states returned by the transformer model to a 2-D tensor
- Test different transformer encoding models: Bert, Albert, XLnet, RoBerta, GPT2.
- Test a TransformerBlock in a classification model
- Test save and load back classification model with transformer encoder block
- Test prepare additional inputs to the TransformerBlock

### Open questions/remarks 

- HuggingFace does not support ragged representation, so the input batch of interaction embeddings is converted to a dense tensor within the TransformerBlock call. 
- In case of masking is being used, How to access the boolean mask tensor?

### CODE REVIEW

File: merlin/models/tf/transformers/transformer_block.py [hunk 1]
```diff
@@ -0,0 +1,548 @@
+#
```
Comment (marcromeyn) [inline merlin/models/tf/transformers/transformer_block.py:1]: Renaming this file to `block.py` is probably more in-line with the rest of our naming

File: merlin/models/tf/transformers/transformer_block.py [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+    Extract serializable custom keras layer `TF*MainLayer` from the HuggingFace model
+    """
+    main_layer = [v for _, v in hf_model.__dict__.items() if isinstance(v, tf.keras.layers.Layer)][
+        0
+    ]
+    return main_layer
+
+
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class PrepareTransformerInputs(tf.keras.layers.Layer):
```
Comment (marcromeyn) [inline merlin/models/tf/transformers/transformer_block.py:53]: Would it make sense to move this to `transforms.py`?

File: merlin/models/tf/transformers/block.py [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+
+
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class TransformerBlock(Block):
+    """
+    Class to support HF Transformers for session-based and sequential-based recommendation models.
+    Parameters
+    ----------
+    transformer: TransformerBody
+        The T4RecConfig, The pre-trained HF model or the custom keras layer TF*MainLayer,
```
Comment (oliverholworthy) [inline merlin/models/tf/transformers/block.py:60]: Is `T4RecConfig` valid here? Or can that part be removed

File: merlin/models/tf/__init__.py [hunk 4]
```diff
@@ -122,6 +122,20 @@
 from merlin.models.tf.prediction_tasks.multi import PredictionTasks
 from merlin.models.tf.prediction_tasks.regression import RegressionTask
 from merlin.models.tf.prediction_tasks.retrieval import ItemRetrievalTask
+from merlin.models.tf.transformers.block import (
```
Comment (oliverholworthy) [inline merlin/models/tf/__init__.py:125]: If the idea is to make transformers an optional dependency. We may not be able to add these imports to this top-level module in this way?

Comment (marcromeyn) [reply merlin/models/tf/__init__.py:125]: Good catch! We should wrap this in a try/catch to handle the case where `transformers` is not installed.

File: tests/unit/tf/transformers/test_block.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    schema = sequence_testing_data.schema.select_by_name(
+        ["item_id_seq", "categories", "user_country"]
+    )
+    schema["user_country"] = schema["user_country"].with_tags(
+        schema["user_country"].tags + "target"
+    )
+    sequence_testing_data.schema = schema
+    dataloader = mm.Loader(sequence_testing_data, batch_size=50)
+    dataloader = dataloader.map(_target_to_onehot)
```
Comment (edknv) [inline tests/unit/tf/transformers/test_block.py:131]: We added an optional parameter `transform` to `Loader` which essentially replaces `.map()` so these two lines can be shortened to
```
dataloader = mm.Loader(sequence_testing_data, batch_size=50, transform=_target_to_onehot)
```

File: merlin/models/tf/transformers/block.py (hunk 3 above)
Comment (sararb) [reply merlin/models/tf/transformers/block.py]: I removed the mention of `T4RecConfig`.

File: tests/unit/tf/transformers/test_block.py (hunk 5 above)
Comment (sararb) [reply tests/unit/tf/transformers/test_block.py]: Ah thank you, I didn't know that option! I have updated the test with your suggestion

File: merlin/models/tf/__init__.py (hunk 4 above)
Comment (sararb) [reply merlin/models/tf/__init__.py]: I added a dependency check `is_transformers_available()` to import the blocks only when `transformers` is installed.

**Review by oliverholworthy — APPROVED** [review]
